### ISSUE TITLE
General question: How do I check that every declared variable has a type definition?

### ISSUE THREAD

**Issue by prettydiff** [typescript-eslint/typescript-eslint#3251]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```TS
let fileIndex:number = 0,
    totalWritten = 0,
    countDir:number = 0,
    statusThrottle:number = Date.now();
```

No type aware linting definitions in my tsconfig.

**Expected Result**
In the above code sample there are 4 variable declarations.  3 are typed while 1 is not.  I am looking for a way to error on that missing type annotation.  I am new to typescript-eslint so I could be missing a basic configuration or a rule.  I have read through the FAQ and applied many of the typscript-eslint rules with success.  I suspect I am just a newb and failed to apply some basic configuration.

Second and largely unrelated question.  Is there is a rule to error on `<Element>myButton` as opposed to `myButton as Element`?

**Actual Result**
The above code sample passes with flying colors.

**Additional Info**
No exceptions from ESLint or this package.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.20.0` |
| `@typescript-eslint/parser`        | `4.20.0` |
| `TypeScript`                       | `4.2.3` |
| `ESLint`                           | `7.23.0` |
| `node`                             | `15.6.0` |

**Comment by bradzacher** [comment]
You should NOT do this.
This is a _bad_ pattern - an anti-pattern.

Each type annotation you provide will ultimately slow TS down. Why? Because it will have to do extra checks to ensure the annotation matches the value.

What do you gain from explicitly annotating every variable's type?
It doesn't make code any clearer because `const x: number = 1` is trivially inferred by a human. If anything it makes code LESS clear as there's now more code to read.
It doesn't make code easier to write as now you're encumbered by annotating every variable. If there are complex types or anonymous object types you now have to explicitly annotate them.

To repeat - you should NOT do this.

**Comment by prettydiff** [comment]
Performance of build times are important.  I spent the day removing many primitive, perhaps over a thousand, type annotations from my 40k LOC project.  There was no performance difference.  It was building between 8.9-9.2 seconds before and after builds at the same speed.  I made the comparison by cloning the project without these changes into a new location and building at each location several times.

Perhaps the performance guidance is based upon older versions of the TypeScript compiler that are no longer necessary.

I am not arguing that this is or isn't an anti-pattern.
